bookkeeper-server/src/main/java/org/apache/bookkeeper/server/service/ScrubberService.java

```diff
@@ -0,0 +1,141 @@
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.server.service;
+
+
+import static org.apache.bookkeeper.bookie.ScrubberStats.DETECTED_FATAL_SCRUB_ERRORS;
+import static org.apache.bookkeeper.bookie.ScrubberStats.DETECTED_SCRUB_ERRORS;
+import static org.apache.bookkeeper.bookie.ScrubberStats.RUN_DURATION;
+
+import com.google.common.util.concurrent.RateLimiter;
+import io.netty.util.concurrent.DefaultThreadFactory;
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.bookkeeper.bookie.ExitCode;
+import org.apache.bookkeeper.bookie.LedgerStorage;
+import org.apache.bookkeeper.server.component.ServerLifecycleComponent;
+import org.apache.bookkeeper.server.conf.BookieConfiguration;
+import org.apache.bookkeeper.stats.Counter;
+import org.apache.bookkeeper.stats.OpStatsLogger;
+import org.apache.bookkeeper.stats.StatsLogger;
+import org.apache.bookkeeper.util.MathUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link org.apache.bookkeeper.common.component.LifecycleComponent} that runs stats provider.
+ */
+public class ScrubberService extends ServerLifecycleComponent {
+    private static final Logger LOG = LoggerFactory.getLogger(ScrubberService.class);
+
+    private static final String NAME = "scrubber";
+    private final ScheduledExecutorService executor;
+    private final Random rng = new Random();
+    private final long scrubPeriod;
+    private final Optional<RateLimiter> scrubRateLimiter;
+    private final AtomicBoolean stop = new AtomicBoolean(false);
+    private final LedgerStorage ledgerStorage;
+
+    private final OpStatsLogger scrubCounter;
+    private final Counter errorCounter;
+    private final Counter fatalErrorCounter;
+
+    public ScrubberService(
+            StatsLogger logger,
+            BookieConfiguration conf,
+            LedgerStorage ledgerStorage) {
+        super(NAME, conf, logger);
+        this.executor = Executors.newSingleThreadScheduledExecutor(
+                new DefaultThreadFactory("ScrubThread"));
+        this.scrubPeriod = conf.getServerConf().getLocalScrubPeriod();
+
+        double rateLimit = conf.getServerConf().getLocalScrubRateLimit();
+        this.scrubRateLimiter = rateLimit == 0 ? Optional.empty() : Optional.of(RateLimiter.create(rateLimit));
+
+        this.ledgerStorage = ledgerStorage;
+
+        this.scrubCounter = logger.getOpStatsLogger(RUN_DURATION);
+        this.errorCounter = logger.getCounter(DETECTED_SCRUB_ERRORS);
+        this.fatalErrorCounter = logger.getCounter(DETECTED_FATAL_SCRUB_ERRORS);
+    }
+
+    private long getNextPeriodMS() {
+        return (long) (((double) scrubPeriod) * (1.5 - rng.nextDouble()) * 1000);
```

Comment:
If scrubPeriod <= 0 this will break

Reply:
I'll update the ServerConfiguration comments to clarify that the service is only enabled if > 0 and add an assert to the constructor to check.

Reply:
The service won't be instantiated otherwise.